src/models/tag.ts

```diff
@@ -0,0 +1,360 @@
+import {
+  SavedItemTagUpdateInput,
+  Tag,
+  SavedItem,
+  TagCreateInput,
+  SaveTagNameConnection,
+  DeleteSavedItemTagsInput,
+  TagUpdateInput,
+  SavedItemTagsInput,
+  DeleteSaveTagResponse,
+} from '../types';
+import config from '../config';
+import { IContext } from '../server/context';
+import { SavedItemDataService, TagDataService } from '../dataService';
+import { UserInputError } from 'apollo-server-express';
+import { ApolloError } from 'apollo-server-errors';
+import { NotFoundError } from '@pocket-tools/apollo-utils';
+import { addslashes } from 'locutus/php/strings';
+import * as Sentry from '@sentry/node';
+
+export class TagModel {
+  private tagService: TagDataService;
+  private saveService: SavedItemDataService;
+  constructor(private context: IContext) {
+    this.saveService = new SavedItemDataService(this.context);
+    this.tagService = new TagDataService(this.context, this.saveService);
+  }
+
+  /**
+   * Convert data layer response to GraphQL Tag entity
+   * Ensures the required fields are there.
+   * This is mostly to catch developer error since the
+   * data model in the db does not match the GraphQL entity
+   */
+  public static toGraphqlEntity(tagResponse: any): Tag {
+    validateTag(tagResponse);
+    return {
+      ...tagResponse,
+      savedItems: tagResponse.savedItems.split(','),
+    };
+  }
+
+  /**
+   * Generate the ID from the DB representation of tag association
+   * Centralizing this logic since it is not modeled in the DB
+   *
+   * @param parent parent Tag entity resolving the ID on
+   * @returns the generated ID for the Tag
+   */
+  public resolveId(parent: Tag): string {
+    return TagModel.encodeId(parent.name);
+  }
+
+  /**
+   * Decode the ID generated from the tag name text
+   * @param id the ID to decode
+   * @returns the tag name text
+   */
+  public static decodeId(id: string): string {
+    const decoded = Buffer.from(id, 'base64').toString();
+    const replace = `${config.data.tagIdSuffix}$`;
+    const regex = new RegExp(replace);
+    return decoded.replace(regex, '');
+  }
+
+  /**
+   * Encode an ID from a tag's name text
+   * @param name the tag name text
+   * @returns the encoded ID
+   */
+  public static encodeId(name: string): string {
+    return Buffer.from(name + config.data.tagIdSuffix).toString('base64');
+  }
+
+  /**
+   * Associate one or more tags to a save
+   */
+  public async createTagSaveConnections(
```

Comment:
trying to understand this workflow:

goes like resolver (calling data models and event emission)-> dataModel (validation, handle business logic and db calls) -> dataService(for db read/write)?

Reply:
yes, but I think the resolver could just be a shallow passthrough in the future and event emission handled in the model? what do you think?

Reply:
ya just like db write, we might have to do the same tag-sanitization before we emit events. in future, we can consolidate them

Reply:
I'll make a ticket so we don't forget. I just didn't want to shove too much into this already big pr 😓